interfaces/hotplug/deviceinfo.go

```diff
@@ -89,3 +90,47 @@ func (h *HotplugDeviceInfo) Attribute(name string) (string, bool) {
 	val, ok := h.Data[name]
 	return val, ok
 }
+
+func (h *HotplugDeviceInfo) firstAttrValueOf(tryAttrs ...string) string {
+	for _, attr := range tryAttrs {
+		if val, _ := h.Attribute(attr); val != "" {
+			return val
+		}
+	}
+	return ""
+}
+
+func (h *HotplugDeviceInfo) String() string {
+	var str []string
+
+	// devname is the name of the device under /dev, eg. /dev/ttyS0;
+	// prefer devname over devpath as this is the one used to talk to the device.
+	if devname := h.DeviceName(); devname != "" {
+		str = append(str, fmt.Sprintf("devname:%s", devname))
+	} else {
+		// devpath is the path of the device under /sys, eg. /sys/devices/pnp0/00:04/tty/ttyS0
+		if devpath := h.DevicePath(); devpath != "" {
+			str = append(str, fmt.Sprintf("devpath:%s", devpath))
+		}
+	}
+	for _, attr := range []string{"MAJOR", "MINOR"} {
+		if val, ok := h.Attribute(attr); ok {
+			// lowercase attribute names as the string representation may appear in task summaries etc.
+			str = append(str, fmt.Sprintf("%s:%s", strings.ToLower(attr), val))
+		}
+	}
+
+	if vendor := h.firstAttrValueOf("ID_VENDOR_FROM_DATABASE", "ID_VENDOR", "ID_VENDOR_ID"); vendor != "" {
+		str = append(str, fmt.Sprintf("vendor:%s", vendor))
+	}
+
+	if model := h.firstAttrValueOf("ID_MODEL_FROM_DATABASE", "ID_MODEL", "ID_MODEL_ID"); model != "" {
+		str = append(str, fmt.Sprintf("model:%s", model))
+	}
+
+	if serial := h.firstAttrValueOf("ID_SERIAL", "ID_SERIAL_SHORT"); serial != "" && serial != "noserial" {
+		str = append(str, fmt.Sprintf("serial:%s", serial))
+	}
+
+	return fmt.Sprintf("<%s>", strings.Join(str, ", "))
```

Comment:
following a bit more the pattern we use for assertions refs I would go for something like:

devname|devpath  (MODEL|VENDOR;  serial: SERIAL)

| means pick the one is defined

And maybe we should establish a maximum length for MODEL,VENDOR,SERIAL and truncate and use ... if they are too long?

Reply:
Sounds good, +1 for truncating. I'd drop serial though per the comment above.